### ISSUE TITLE
Allow warnings when a colon isn't followed by a space

### ISSUE THREAD

**Issue by beyarkay** [adrienverge/yamllint#563]
I recently got bit by this piece of valid-but-incorrect YAML:

```yaml
- my_object: some_value
  source: https://www.google.com
  areas:
  - {name: value, other_key: []}
  - {name:value,other_key: []}         # <-- this line in particular
  - {name:name, other_key: []}
```

because I didn't have a space after the colon for the `name` key, YAML parses this perfectly fine. I realise that requiring a space after a colon isn't always desirable becuase sometimes you want an actual colon (like in `https://www.google.com`), but is there an option which would warn about non-quoted strings with potentially confusing special characters?

For example, these would cause warnings: `{name:value}`, `https://www.google.com` because what they mean to YAML might not be what the user intended, and would recommend either quoting the special character: `{"name:value"}`, `"https://www.google.com"`, or using a space: `{name: value}`, `https: //www.google.com`

See also [this Stack overflow question](https://stackoverflow.com/questions/42124227/why-does-the-yaml-spec-mandate-a-space-after-the-colon), which conveys that this is a common misconception.

**Comment by adrienverge** [comment]
Hello, and thanks for this interesting thought!

The short answer is: no, you cannot currently check this with yamllint. The only check I can imagine is a rule that reports errors for any unquoted string that contains `:`. For instance:

```yaml
https://www.website.com/
# error: should be written:
https: //[www.website.com](http://www.website.com/)
# or:
"https://www.website.com/"
```

But in my opinion:
- This defeats the usefulness of YAML: being able to write `ke:y: value` if you want to. Or simply write URLs without quoting them (like you exampled).
- This reaches the limit of what a linter can do: check the content / the actual data (here, the string in `{"name:value,other_key": []}`.
- There are probably many other similar situations if you miss characters in YAML (e.g. `[list, of, 3,items,because,missing,spaces]`): a linter cannot guess them all.

**Comment by beyarkay** [comment]
Hmm okay. I understand where you're coming from, although I feel like having this as an off-by-default option could dramatically improve certain use-cases.

For me, I'm currently looking at writing my own schema-checker, because I can't find anything out there where I can specify the schema I expect my YAML to have, and then for the tool to validate it for me. (if you know of anything I'd be super excited to check it out though!). However, if I could do some basic validation with the linter (such as checking for special characters in keys/values) then I probably could get away without the schema-validator.

I do agree that "being explicit to avoid ambiguity" kinda goes contrary to why YAML is often better than JSON. I guess I was hoping that I'd be able to have my cake and eat it too 🥲. But I totally see your viewpoint, and I see that this could very easily be out of scope for a linter.

I don't suppose you know of other YAML projects which try to do more than just linting?

**Comment by adrienverge** [comment]
I couldn't summarize things better :+1: 

I'm not aware of any tool to validate YAML *content / data*, but if I had to do it I would make a script that:
1. Turns YAML into JSON.
   e.g. `python -c 'import sys, json, yaml; print(json.dumps(yaml.safe_load(sys.stdin)))' < file.yaml`
3. Enforce a [JSON Schema](https://json-schema.org/).

**Comment by beyarkay** [comment]
Pity that YAML would have to depend on JSON, but I guess that's where we're at. 

Thanks by the way for communicating so openly and objectively! Some maintainers can get really annoyed when bug reports are opened.

I'll see what I can do, but I'm not sure if there's something nice and easy out there for me (yet...).
